server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java

```diff
@@ -789,15 +789,20 @@ private CustomSettableFuture(
 
     public void resolve()
     {
-      synchronized (statusRefs) {
+      synchronized (requestStatusesLock) {
```

Comment:
Thanks for calling it out. Yes, it is better if I just add a comment explaining the logic.

I made the switch to make sure that the update invoked by `requestStatuses.invalidate()` is safeguarded.
The current synchronization on `statusRefs` essentially meant that only one thread can be executing the `resolve` of a given `CustomSettableFuture` at any point. That condition still holds but with less concurrency, i.e. in the new code, only one thread can be executing the `resolve` of _any_ `CustomSettableFuture` at any point.

So, in other words, while multiple segments might be loaded by the historical in parallel, the futures will only be resolved serially.

This will not have any other impact on the correctness or performance of this code.

Reply:
Thanks for the explanation. It looks like there is no longer anything synchronizing on `statusRefs`, so the new code is somewhat simpler. Please do add a comment to `requestStatusesLock`, or use `@GuardedBy` on the things it is meant to guard, as appropriate, in order to make the logic easier to follow for others.

Reply:
Added an `@GuardedBy` for `requestStatuses` and some comments.